File: src/core/sync/excluded-keys.ts
/**
 * excludeKeys 的「读写两侧」语义（docs/m2-plan.md §2.5 / DESIGN §2.2、§2.7；W4 唯一 owner）。
 *
 * 三重语义在这里收口两处，第三处（植回）由 plan.ts 在合并完成后调用 `plantExcludedKeys`：
 *   1. **push 写 store 前**：`applyExcludeKeyPlaceholders` 把 excludeKeys 列出的顶层键值整体替换为
 *      `__REQUIRED__`，使密钥值永不进入 git。
 *   2. **pull / merge 判定前**：`stripSnapshotExcludeKeys` 按 category 剥离这些顶层键 —— 直接复用
 *      engine 冻结的 `stripExcludeKeys`，与 `homer status`（`src/cli/render.ts` 的
 *      `stripAdapterExcludedKeys`）**同一实现点**。于是 store 里的 `__REQUIRED__` 占位符既不会被
 *      当成「远端变更」，也不会流入工具目录。
 *   3. **merge 完成后**：`plantExcludedKeys` 把 local 原文件中的 excluded 键值植回 merged ——
 *      本地密钥永不被远端覆盖；local 缺该键 → 不植回（命令层据此记 warning「缺失必填项」）。
 *
 * 纯函数、零 fs。依赖面刻意收窄为 `types.ts`（纯类型）+ engine 的 `stripExcludeKeys`：
 * 连 `entry-kind.ts` 也不 import（`isJsonObject` 就地实现），保证 W4 只碰 plan §3-P1-W4 授权的依赖。
 */

import { stripExcludeKeys } from '../engine/index.js';
import type { AdapterSnapshot, CategorySnapshot, HomerConfig, SnapshotEntry, SnapshotFiles } from '../types.js';

/** excludeKeys 占位符字面量（DESIGN §2.2：参考 legout/pi-config 的 `__REQUIRED__` 方案）。 */
export const REQUIRED_PLACEHOLDER = '__REQUIRED__';

/** 冻结的确定性 JSON 落盘格式（plan §2.5：`JSON.stringify(merged, null, 2) + '\n'`）。 */
export function serializeJsonContent(value: unknown): string {
  return `${JSON.stringify(value, null, 2)}\n`;
}

/**
 * 非 null、非数组的对象（JSON 对象语义）。
 * 与 `core/entry-kind.ts` 的 `isPlainObject` 同义；此处就地实现是为了不扩大 W4 的依赖面
 * （plan §3-P1-W4 限定「只 import types / engine / sync-types」）。
 */
export function isJsonObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/** 可被 `JSON.parse` 接受的文本 → 值；失败 → `undefined`。 */
export function parseJsonContent(content: string): unknown {
  try {
    return JSON.parse(content) as unknown;
  } catch {
    return undefined;
  }
}

/**
 * `adapterId` → `{ categoryName → excludeKeys }`，只保留非空条目。
 * 与 `homer status` 的剥离口径同构（planPull / checkPushSafety 共用本函数）。
 */
export function excludeKeysByCategory(config: HomerConfig, adapterId: string): Record<string, string[]> {
  const adapter = config.adapters[adapterId];
  if (adapter === undefined) return {};

  const out: Record<string, string[]> = {};
  for (const [name, category] of Object.entries(adapter.categories)) {
    const keys = category.excludeKeys;
    if (keys !== undefined && keys.length > 0) out[name] = keys;
  }
  return out;
}

const EMPTY_KEYS: string[] = [];

/** 单个 category 的 excludeKeys（未配置 → 空数组，零分配）。 */
export function excludedKeysFor(config: HomerConfig, adapterId: string, category: string): readonly string[] {
  return config.adapters[adapterId]?.categories[category]?.excludeKeys ?? EMPTY_KEYS;
}

/** 按 config 剥离单快照的 excludeKeys（判定前置口径；无配置 → 原快照返回）。 */
export function stripSnapshotExcludeKeys(snapshot: AdapterSnapshot, config: HomerConfig): AdapterSnapshot {
  const byCategory = excludeKeysByCategory(config, snapshot.adapterId);
  if (Object.keys(byCategory).length === 0) return snapshot;
  return stripExcludeKeys(snapshot, byCategory);
}

/**
 * push 写 store 前：excludeKeys 顶层键 → 值 `__REQUIRED__`。
 *
 * 只替换**原本就存在**的键：不凭空造出用户从未拥有的配置键（「缺失必填项」由 pull 侧
 * 「local 缺该键 → 不植回」的反向校验负责）。文件中没有命中任何 excluded 键 → 条目字节不变，
 * 避免给 store 引入无意义的格式 diff。
 */
export function applyExcludeKeyPlaceholders(snapshot: AdapterSnapshot, config: HomerConfig): AdapterSnapshot {
  const byCategory = excludeKeysByCategory(config, snapshot.adapterId);
  return {
    adapterId: snapshot.adapterId,
    categories: snapshot.categories.map((category) => {
      const keys = byCategory[category.category];
      if (keys === undefined) return cloneCategory(category);
      return { ...category, files: placeholderFiles(category.files, keys) };
    }),
  };
}

/**
 * merge 完成后把 local 原文件中的 excluded 键值植回 merged（本地永不被远端覆盖）。
 * local 缺该键 → 不植回（保留 merged 中「无此键」的事实）；无可植回键 → 原值返回（零拷贝）。
 */
export function plantExcludedKeys(
  merged: Record<string, unknown>,
  localValue: unknown,
  keys: readonly string[],
): Record<string, unknown> {
  if (keys.length === 0 || !isJsonObject(localValue)) return merged;

  const planted: [string, unknown][] = [];
  let changed = false;
  for (const [key, value] of Object.entries(merged)) {
    if (keys.includes(key) && hasOwn(localValue, key)) {
      planted.push([key, localValue[key]]); // 本地值优先（防御：正常流程下 merged 已被 strip）
      changed = true;
      continue;
    }
    planted.push([key, value]);
  }
  for (const key of keys) {
    if (hasOwn(merged, key) || !hasOwn(localValue, key)) continue;
    planted.push([key, localValue[key]]);
    changed = true;
  }

  // Object.fromEntries 用 CreateDataProperty 落键（`__proto__` 当普通数据键，不污染原型）。
  return changed ? Object.fromEntries(planted) : merged;
}

/* ------------------------------------------------------------------ */
/* 内部实现                                                            */
/* ------------------------------------------------------------------ */

function cloneCategory(category: CategorySnapshot): CategorySnapshot {
  return { ...category, files: new Map(category.files) };
}

function hasOwn(value: object, key: string): boolean {
  // 原型键（`__proto__` / `constructor`）不得沿原型链取到值。
  return Object.prototype.hasOwnProperty.call(value, key);
}

function placeholderFiles(files: SnapshotFiles, keys: readonly string[]): SnapshotFiles {
  const out: SnapshotFiles = new Map();
  for (const [relPath, entry] of files) out.set(relPath, placeholderEntry(entry, keys));
  return out;
}

function placeholderEntry(entry: SnapshotEntry, keys: readonly string[]): SnapshotEntry {
  if (entry.kind !== 'json') return entry;
  const value = parseJsonContent(entry.content);
  if (!isJsonObject(value)) return entry;
  if (!Object.keys(value).some((key) => keys.includes(key))) return entry;

  const replaced = Object.entries(value).map(
    ([key, item]) => [key, keys.includes(key) ? REQUIRED_PLACEHOLDER : item] as const,
  );
  return { kind: 'json', content: serializeJsonContent(Object.fromEntries(replaced)) };
}

File: internal/sync/excludedkeys.go
package sync

import (
	"sort"
	"strings"

	"github.com/zzjcool/homer-cli/internal/core"
	"github.com/zzjcool/homer-cli/internal/orderedjson"
)

// RequiredPlaceholder is the value stored in git for an excluded local key.
// It is deliberately a plain sentinel rather than a secret-bearing value.
const RequiredPlaceholder = "__REQUIRED__"

// REQUIRED_PLACEHOLDER preserves the migration-facing spelling used by the
// TypeScript implementation and its fixture names.
const REQUIRED_PLACEHOLDER = RequiredPlaceholder

// ParseJSONContent parses a snapshot entry without exposing parser errors to
// plan callers. A false result means malformed JSON.
func ParseJSONContent(content string) (orderedjson.Value, bool) {
	value, err := orderedjson.Parse([]byte(content))
	return value, err == nil
}

func parseJSONContent(content string) (orderedjson.Value, bool) {
	return ParseJSONContent(content)
}

func parseJsonContent(content string) (orderedjson.Value, bool) {
	return ParseJSONContent(content)
}

// IsJSONObject is the shared JSON-object predicate for merge eligibility and
// excluded-key operations. Arrays, scalars, null, and malformed input are not
// object records.
func IsJSONObject(value orderedjson.Value) bool {
	object, ok := value.(*orderedjson.Object)
	return ok && object != nil
}

func isJSONObject(value orderedjson.Value) bool { return IsJSONObject(value) }
func isJsonObject(value orderedjson.Value) bool { return IsJSONObject(value) }

// IsJsonObject is a spelling-compatible alias for callers using Go's initialism
// convention less strictly.
func IsJsonObject(value orderedjson.Value) bool { return IsJSONObject(value) }

// SerializeJSONContent is the canonical on-disk JSON format for merged files
// and placeholder snapshots: JSON.stringify(value, null, 2) plus one newline.
func SerializeJSONContent(value orderedjson.Value) string {
	return string(orderedjson.SerializeFile(value))
}

func serializeJSONContent(value orderedjson.Value) string { return SerializeJSONContent(value) }
func serializeJsonContent(value orderedjson.Value) string { return SerializeJSONContent(value) }

// ExcludedKeysByCategory returns configured top-level excluded keys by
// category. The returned slices are copies so callers cannot mutate config.
func ExcludedKeysByCategory(config core.HomerConfig, adapterID string) map[string][]string {
	adapter, ok := config.Adapters[adapterID]
	if !ok {
		return map[string][]string{}
	}
	out := make(map[string][]string)
	for category, cfg := range adapter.Categories {
		if len(cfg.ExcludeKeys) == 0 {
			continue
		}
		out[category] = append([]string(nil), cfg.ExcludeKeys...)
	}
	return out
}

func excludedKeysByCategory(config core.HomerConfig, adapterID string) map[string][]string {
	return ExcludedKeysByCategory(config, adapterID)
}

// ExcludedKeysFor returns the configured excluded top-level keys for one
// adapter/category. No keys is represented by an empty slice.
func ExcludedKeysFor(config core.HomerConfig, adapterID, category string) []string {
	adapter, ok := config.Adapters[adapterID]
	if !ok {
		return nil
	}
	return append([]string(nil), adapter.Categories[category].ExcludeKeys...)
}

func excludedKeysFor(config core.HomerConfig, adapterID, category string) []string {
	return ExcludedKeysFor(config, adapterID, category)
}

// StripSnapshotExcludeKeys removes configured top-level keys from JSON entries
// for judgement. Like the TypeScript JSON.stringify path, valid entries are
// compactly serialized; raw entries are never mutated.
func StripSnapshotExcludeKeys(snapshot core.AdapterSnapshot, config core.HomerConfig) core.AdapterSnapshot {
	byCategory := ExcludedKeysByCategory(config, snapshot.AdapterID)
	out := cloneSnapshot(snapshot)
	if len(byCategory) == 0 {
		return out
	}
	for index := range out.Categories {
		category := &out.Categories[index]
		keys := byCategory[category.Category]
		if len(keys) == 0 {
			continue
		}
		files := make(core.SnapshotFiles, len(category.Files))
		for path, entry := range category.Files {
			files[path] = stripEntry(entry, keys)
		}
		category.Files = files
	}
	return out
}

func stripSnapshotExcludeKeys(snapshot core.AdapterSnapshot, config core.HomerConfig) core.AdapterSnapshot {
	return StripSnapshotExcludeKeys(snapshot, config)
}

// ApplyExcludeKeyPlaceholders replaces only excluded keys that actually exist
// in a local snapshot. It never invents a missing required key.
func ApplyExcludeKeyPlaceholders(snapshot core.AdapterSnapshot, config core.HomerConfig) core.AdapterSnapshot {
	byCategory := ExcludedKeysByCategory(config, snapshot.AdapterID)
	out := cloneSnapshot(snapshot)
	for index := range out.Categories {
		category := &out.Categories[index]
		keys := byCategory[category.Category]
		if len(keys) == 0 {
			continue
		}
		files := make(core.SnapshotFiles, len(category.Files))
		for path, entry := range category.Files {
			files[path] = placeholderEntry(entry, keys)
		}
		category.Files = files
	}
	return out
}

func applyExcludeKeyPlaceholders(snapshot core.AdapterSnapshot, config core.HomerConfig) core.AdapterSnapshot {
	return ApplyExcludeKeyPlaceholders(snapshot, config)
}

// PlantExcludedKeys puts local values back into a successful merged object.
// Existing local keys win; a key absent from local is deliberately not
// reintroduced. The returned value is a new object only when a planting was
// needed.
func PlantExcludedKeys(merged, localValue orderedjson.Value, keys []string) orderedjson.Value {
	mergedObject, ok := merged.(*orderedjson.Object)
	if !ok || mergedObject == nil || !IsJSONObject(localValue) || len(keys) == 0 {
		return merged
	}
	localObject := localValue.(*orderedjson.Object)
	keySet := make(map[string]struct{}, len(keys))
	for _, key := range keys {
		keySet[key] = struct{}{}
	}

	out := &orderedjson.Object{Keys: make([]string, 0, len(mergedObject.M)+len(keys)), M: make(map[string]orderedjson.Value)}
	changed := false
	for _, key := range objectKeys(mergedObject) {
		value := mergedObject.M[key]
		if _, excluded := keySet[key]; excluded {
			if local, exists := localObject.M[key]; exists {
				value = local
				changed = true
			}
		}
		out.Keys = append(out.Keys, key)
		out.M[key] = value
	}
	for _, key := range keys {
		if _, exists := mergedObject.M[key]; exists {
			continue
		}
		if local, exists := localObject.M[key]; exists {
			out.Keys = append(out.Keys, key)
			out.M[key] = local
			changed = true
		}
	}
	if !changed {
		return merged
	}
	return out
}

func plantExcludedKeys(merged, localValue orderedjson.Value, keys []string) orderedjson.Value {
	return PlantExcludedKeys(merged, localValue, keys)
}

func cloneSnapshot(snapshot core.AdapterSnapshot) core.AdapterSnapshot {
	out := snapshot
	out.Categories = make([]core.CategorySnapshot, len(snapshot.Categories))
	for index, category := range snapshot.Categories {
		out.Categories[index] = category
		out.Categories[index].Files = make(core.SnapshotFiles, len(category.Files))
		for path, entry := range category.Files {
			out.Categories[index].Files[path] = entry
		}
	}
	return out
}

func stripEntry(entry core.SnapshotEntry, keys []string) core.SnapshotEntry {
	if entry.Kind != "json" {
		return entry
	}
	value, ok := ParseJSONContent(entry.Content)
	if !ok {
		return entry
	}
	stripped := orderedjson.StripTopKeys(value, keys)
	return core.SnapshotEntry{Kind: "json", Content: compactSerialize(stripped)}
}

func placeholderEntry(entry core.SnapshotEntry, keys []string) core.SnapshotEntry {
	if entry.Kind != "json" {
		return entry
	}
	value, ok := ParseJSONContent(entry.Content)
	if !ok || !IsJSONObject(value) {
		return entry
	}
	object := value.(*orderedjson.Object)
	keySet := make(map[string]struct{}, len(keys))
	for _, key := range keys {
		keySet[key] = struct{}{}
	}
	matched := false
	for key := range object.M {
		if _, ok := keySet[key]; ok {
			matched = true
			break
		}
	}
	if !matched {
		return entry
	}

	replaced := &orderedjson.Object{Keys: make([]string, 0, len(object.M)), M: make(map[string]orderedjson.Value, len(object.M))}
	for _, key := range objectKeys(object) {
		value := object.M[key]
		if _, replace := keySet[key]; replace {
			value = RequiredPlaceholder
		}
		replaced.Keys = append(replaced.Keys, key)
		replaced.M[key] = value
	}
	return core.SnapshotEntry{Kind: "json", Content: SerializeJSONContent(replaced)}
}

func objectKeys(object *orderedjson.Object) []string {
	if object == nil {
		return nil
	}
	seen := make(map[string]struct{}, len(object.M))
	keys := make([]string, 0, len(object.M))
	for _, key := range object.Keys {
		if _, exists := object.M[key]; !exists {
			continue
		}
		if _, exists := seen[key]; exists {
			continue
		}
		seen[key] = struct{}{}
		keys = append(keys, key)
	}
	extra := make([]string, 0)
	for key := range object.M {
		if _, exists := seen[key]; !exists {
			extra = append(extra, key)
		}
	}
	sort.Strings(extra)
	return append(keys, extra...)
}

// compactSerialize minifies orderedjson's canonical output without touching
// whitespace inside JSON strings. This preserves insertion order while
// matching JSON.stringify(value) for the judgement snapshots.
func compactSerialize(value orderedjson.Value) string {
	pretty := orderedjson.Serialize(value)
	var out strings.Builder
	out.Grow(len(pretty))
	inString := false
	escaped := false
	for _, char := range string(pretty) {
		if inString {
			out.WriteRune(char)
			if escaped {
				escaped = false
			} else if char == '\\' {
				escaped = true
			} else if char == '"' {
				inString = false
			}
			continue
		}
		if char == '"' {
			inString = true
			out.WriteRune(char)
			continue
		}
		switch char {
		case ' ', '\t', '\n', '\r':
			continue
		default:
			out.WriteRune(char)
		}
	}
	return out.String()
}
